File: 532.K-diff-Pairs-in-an-Array.go
// https://leetcode.com/problems/k-diff-pairs-in-an-array/description/
//
// algorithms
// Easy (28.46%)
// Total Accepted:    46.7k
// Total Submissions: 164.1k
// beats 54.84% of golang submissions

package leetcode

func findPairs(nums []int, k int) int {
	if k < 0 {
		return 0
	}

	numMap := make(map[int]int)

	for _, num := range nums {
		if cntNum, ok := numMap[num]; ok {
			numMap[num] = cntNum + 1
		} else {
			numMap[num] = 1
		}
	}

	res := 0
	for key, value := range numMap {
		if k == 0 {
			if value > 1 {
				res++
			}
			continue
		}
		if _, ok := numMap[key+k]; ok {
			res++
		}
	}

	return res
}

File: 532.K-diff-Pairs-in-an-Array.java
// https://leetcode.com/problems/convert-bst-to-greater-tree/
//
// algorithms
// Easy (30.4%)
// Total Accepted: 75,041
// Total Submissions: 246,822
// beats 100.0% of java submissions


class Solution {
    public int findPairs(int[] nums, int k) {
        if (k < 0) {
            return 0;
        }
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int n : nums) {
            map.put(n, map.getOrDefault(n, 0) + 1);
        }

        int res = 0;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            int n = entry.getKey();
            int v = entry.getValue();

            if (k == 0) {
                if (v > 1) {
                    res++;
                }
            } else if (map.containsKey(n + k)) {
                res++;
            }
        }

        return res;
    }
}